cookbook/controller/argument_value_resolver.rst

```diff
@@ -0,0 +1,163 @@
+
+  * The :class:`Symfony\\Component\\HttpKernel\\Controller\\ArgumentValueResolver\\VariadicValueResolver`
+    verifies in the request if your data is an array and will add all of them to the argument list.
+    When the action is called, the last (variadic) argument will contain all the values of this array.
+
+.. note::
+
+    In older versions of Symfony this logic was all resolved within the ``ControllerResolver``. The
+    old functionality is moved to the ``LegacyArgumentResolver``, which contains the previously
+    used resolving logic.
+
+Adding a New Value Resolver
+---------------------------
+
+Adding a new value resolver requires one class and one service defintion. In our next example, we
+will be creating a shortcut to inject the ``User`` object from our security. Given you write the following
+action::
+
+    namespace AppBundle\Controller;
+
+    use AppBundle\User;
+    use Symfony\Component\HttpFoundation\Response;
+
+    class UserController
+    {
+        public function indexAction(User $user)
+        {
+            return new Response('<html><body>Hello '.$user->getUsername().'!</body></html>');
+        }
+    }
+
+Somehow you will have to get the ``User`` object and inject it into our action. This can be done
+by implementing the :class:`Symfony\\Component\\HttpKernel\\Controller\\ArgumentValueResolverInterface`.
+This interface specifies that you have to implement two methods::
+
+    interface ArgumentValueResolverInterface
+    {
+        public function supports(Request $request, ArgumentMetadata $argument);
+        public function resolve(Request $request, ArgumentMetadata $argument);
+    }
+
+  * The ``supports()`` method is used to check whether the resolver supports the given argument. It will
+    only continue if it returns ``true``.
+
+  * The ``resolve()`` method will be used to resolve the actual value just acknowledged by
+    ``supports()``. Once a value is resolved you can ``yield`` the value to the ``ArgumentResolver``.
+
+  * The ``Request`` object is the current ``Request`` which would also be injected into your
+    action in the forementioned functionality.
+
+  * The :class:``Symfony\\Component\\HttpKernel\\ControllerMetadata\\ArgumentMetadata`` represents
+    information retrieved from the method signature for the current argument it's trying to resolve.
+
+.. note::
+
+    The ``ArgumentMetadata`` is a simple data container created by the
+    :class:``Symfony\\Component\\HttpKernel\\ControllerMetadata\\ArgumentMetadataFactory``. This
+    factory will work on every supported PHP version but might give different results. E.g. the
+    ``isVariadic()`` will never return true on PHP 5.5 and only on PHP 7.0 and higher it will give
+    you basic types when calling ``getType()``.
+
+Now that you know what to do, you can implement this interface. In order to get the current ``User``,
+you will have to get it from the ``TokenInterface`` which is in the ``TokenStorageInterface``::
```

Comment:
Let's split this long sentence in a couple smaller ones to make it easier to read: "To get the current `User`, you need the current security token. This token can be retrieved from the current token storage."